scripts/utils/options.ts

```diff
@@ -0,0 +1,279 @@
+/**
+ * Use commander and prompts to gather a list of options for a script
+ */
+
+import prompts, { Falsy, PrevCaller, PromptType } from 'prompts';
+import type { PromptObject } from 'prompts';
+import program from 'commander';
+import type { Command } from 'commander';
+import kebabCase from 'lodash/kebabCase';
+
+// Option types
+
+export type OptionId = string;
+export type BaseOption = {
+  description?: string;
+  /**
+   * By default the one-char version of the option key will be used as short flag. Override here,
+   *   e.g. `shortFlag: 'c'`
+   */
+  shortFlag?: string;
+  /**
+   * What type of prompt to use? (return false to skip, true for default)
+   */
+  promptType?: PromptType | Falsy | PrevCaller<string, PromptType | boolean>;
+};
+
+export type BooleanOption = BaseOption & {
+  /**
+   * Does this option default true?
+   */
+  inverse?: boolean;
+};
+
+export type StringOption = BaseOption & {
+  /**
+   * What values are allowed for this option?
+   */
+  values: string[];
+  /**
+   * Is a value required for this option?
+   */
+  required?: boolean;
+};
+
+export type StringArrayOption = BaseOption & {
+  /**
+   * What values are allowed for this option?
+   */
+  values: string[];
+  /**
+   * This must be set to true
+   */
+  multiple: true;
+};
+
+// StringArrayOption requires `multiple: true;` but unless you use `as const` an object with
+// { multiple: true } will be inferred as `multiple: boolean;`
+type StringArrayOptionMatch = Omit<StringArrayOption, 'multiple'> & { multiple: boolean };
```

Comment:
Sadly, to have it inferred, you either need a contextual type or `as const` indeed. You don't have to do it on the object, you can also do it on `multiple: true as const`, which seems fine to me.


A contextual type like `satisfies OptionSpecifier` in options.test.ts may be ideal, but only in a future version of TS 😅 . Otherwise it needs a hard annotation `const allOptions: OptionSpecifier` or the contextual type can be provided using a factory function.

Reply:
I'm pretty open to whatever folks prefer here, so I will definitely be guided by you here @kasperpeulen. What's interesting about how it's done here is that in the source file you don't have to do anything special apart from pass the options in as a raw object (i.e you don't need to type it or use `as const`): https://github.com/storybookjs/storybook/blob/457714dd06a3201fc7d792bcfc986555b9a3a0ab/scripts/example.ts#L15-L58

That seems good because there's less possibility of messing it up. Then again, with this stuff what does "messing it up" mean? It means that the object you get back out of the call (to say `getOptions`) will have the wrong type. Which I guess will be pretty obvious, and pretty quickly fixed.

---

The other thing I am wondering folks' option on (cc @yannbf etc) is the use of these fields to imply the type of the option. i.e. using:

```ts
const booleanOption = {};
const stringOption = { values: ['a'] }; // the presence of `value` implies this is a string type.
```

As opposed to:

```ts
const booleanOption = { type: 'boolean' as const };
const stringOption = { type: 'string' as const, values: ['a'] };
```

here it is much more explicit, but also wordier.

I guess the major downside of the current "implied" approach is if say down the track we want to get rid of the restriction that `StringOption`s must have an array of values, we might have an issue. But then again we can always address that if/when it happens (it wouldn't be that hard to update the codebase and add a `type` field to all the options).

Reply:
I think that the `true as const` on `multiple` is only needed in places where there is no contextual type, so for example here:

https://github.com/storybookjs/storybook/blob/93f3e0e5b5e7177440ebae3d514147fe83d50c6d/scripts/utils/options.test.ts#L12-L30

But in example.ts it should not be needed, as the arguments of the function call are contextually typed:

https://github.com/storybookjs/storybook/blob/93f3e0e5b5e7177440ebae3d514147fe83d50c6d/scripts/example.ts#L18-L29

I checked that with or without the `as const`, the type inferred is properly, and the `as string[]` here is not needed anymore 😄 :

https://github.com/storybookjs/storybook/blob/93f3e0e5b5e7177440ebae3d514147fe83d50c6d/scripts/example.ts#L180-L184

Reply:
@yannbf @tmeasday 

About the structure implying the type. I think it is considered best practice to always have a clear discriminant in unions (they are also often called "discriminated unions", "tagged unions").  But I don't know, because of TS flexbility it is more and more pushed in directions that is quite unexplored in other languages with those kind of features 😅.

One downside of this approach is that you need type predicates to keep it readable:

https://github.com/storybookjs/storybook/blob/93f3e0e5b5e7177440ebae3d514147fe83d50c6d/scripts/utils/options.ts#L79-L89

And although they are pretty awesome, they are also not typesafe, in the sense that TS won't check at all if you predicate is correct. 

You could make a discriminant `type?:  'string' | 'array'` and `undefined` implying it being a boolean, to make it shorter.

I was also thinking that, if this features grows, you might consider using `zod` for this. It is excellent in specifying runtime types, inferring the TS type from that, and also adding extra validation on top of that. It is often used in combination with form libraries, and I guess this is pretty much a form 😅.

It could look like this:

```ts
async function getOptions() {
  return getOptionsOrPrompt('yarn example', {
    framework: {
      type: z.enum(['react', 'angular']),
      description: 'Which framework would you like to use?',
    },
    addon: {
      type: z.array(z.enum(['a11y', 'storysource'])),
      description: 'Which extra addons (beyond the CLI defaults) would you like installed?',
    },
    create: {
      // default type may be z.boolean.default(false)
      description: 'Create the example from scratch (rather than degitting it)?',
    },
    start: {
      // inversed
      type: z.boolean().default(true),
      description: 'Start the example Storybook?',
    },
    email: {
      // all kind of helpers for free
      type: z.string().email(),
      description: 'Age',
    },
    age: {
      type: z.number().int().min(0),
      description: 'Age',
    },
```

Reply:
Yeah, cool. I think I'll stop swimming upstream and just use the discriminated union here.

> I was also thinking that, if this features grows, you might consider using zod for this

So yeah, I think `zod` is definitely conceptually in the same headspace as what I was doing here (love "parse, don't validate"!).

I guess the tricky part here is that zod is a parser/validator but what we want to parse/validate is CLI flags and commander is already doing that for us. So I'm not exactly sure how easy it will be to extract all the information we need to give to commander from zod. I'll have a play with it! It's definitely better than coming up with some new API for `.string().email()` etc ;)

Reply:
@kasperpeulen one quick question about `zod` -- if we were going to do `{ type: z.something() }` how would we then do runtime branching based on zod type? For instance, to pick the correct `prompt` to show the user if they haven't filled out an option? 

Previously we used the `isStringOption()` and friends to do that, is there a way to do that, considering that things like `z.string().default()` can end up with multiple layers of wrapped validators?

Reply:
@tmeasday 

Yes, I do love zod parse don't validate as well :D I like to make sure all incoming data is handled at the very edges of the program, so that you can really trust TS inside the application, parse it right away and fail fast if needed.

I think it should be possible to get all the information needed for commander and prompt with runtime branching. You can unwrap the default by accessing `_def.innerType`. Possibly, recursively, but to start we could maybe constrain the type a bit, like this:

```ts
type OptionType =
  | z.ZodDefault<z.ZodBoolean | z.ZodEnum<[string, ...string[]]>>
  | z.ZodBoolean
  | z.ZodEnum<[string, ...string[]]>
  | z.ZodArray<z.ZodEnum<[string, ...string[]]>>;

function toPromptType(option: {
  key: string;
  type: OptionType;
  description: string;
}): PromptObject {
  const { type, key, description } = option;
  if (type instanceof z.ZodDefault) {
    return {
      ...toPromptType({ ...option, type: type._def.innerType }),
      initial: type._def.defaultValue(),
    };
  }
  if (type instanceof z.ZodBoolean) {
    return { type: 'toggle', name: key, message: description };
  }
  if (type instanceof z.ZodArray) {
    return {
      name: key,
      type: 'autocompleteMultiselect',
      choices: type.element.options.map((it) => ({ title: it, value: it })),
    };
  }
  return {
    name: key,
    type: 'select',
    choices: type.options.map((it) => ({ title: it, value: it })),
  };
}
```

Reply:
Ok, got it! I thought that was probably a way to go but I guess I was hoping there was something simpler I was missing ;)

Reply:
Hehe, yeah, it is certainly not zod biggest strength. But you do get type inference for free with this approach (which is a strength of zod):

<img width="679" alt="image" src="https://user-images.githubusercontent.com/1035299/183306463-95912018-b988-4546-8d84-d2ee97981789.png">

Btw, zod only works well when the tsconfig is set to strict: true. I couldn't get the infer to work if I didn't change the tsconfig.

Reply:
Yes that's nice although I got something similar going last night in our homegrown system too :
<img width="584" alt="image (1)" src="https://user-images.githubusercontent.com/132554/183316082-fbee6c5c-30bf-4871-8dec-859b0a20a9db.png">
)
<img width="639" alt="image" src="https://user-images.githubusercontent.com/132554/183316085-8b9f7dd8-eac9-4f3f-ab79-4915a0aa1907.png">